runtime/src/iree/io/formats/safetensors/safetensors_format.c

```diff
@@ -0,0 +1,520 @@
+  // offset that all data ranges are relative to. Verifies that the header and
+  // base offset is in range but each entry data range still needs to be
+  // verified.
+  uint64_t remaining_bytes = file_contents.data_length;
+  uint64_t header_length = 0;
+  if (remaining_bytes < sizeof(header_length)) {
+    return iree_make_status(IREE_STATUS_OUT_OF_RANGE,
+                            "insufficient capacity for safetensors header "
+                            "length (need at least %" PRIhsz
+                            " bytes but have %" PRIu64 ")",
+                            sizeof(header_length), remaining_bytes);
+  }
+  header_length =
+      iree_unaligned_load_le_u64((const uint64_t*)file_contents.data);
+  remaining_bytes -= sizeof(header_length);
+  if (remaining_bytes < header_length) {
+    return iree_make_status(IREE_STATUS_OUT_OF_RANGE,
+                            "insufficient capacity for safetensors header "
+                            "contents (declared as %" PRIu64
+                            " but only %" PRIu64 " bytes available)",
+                            header_length, remaining_bytes);
+  }
+  const iree_string_view_t header_json = iree_make_string_view(
+      (const char*)file_contents.data + sizeof(header_length),
+      (iree_host_size_t)header_length);
+  const uint64_t base_offset = sizeof(header_length) + header_length;
+  remaining_bytes -= header_length;
+
+  // Parses a safetensors |header_json| blob and emits entries to |index|.
+  // Each entry is bounds checked against the |data_size| of the file (bytes
+  // excluding the header, relative to |base_offset|).
+  iree_io_enumerate_safetensors_entry_state_t enumerate_state = {
+      .file_handle = file_handle,
+      .base_offset = base_offset,
+      .data_size = remaining_bytes,
+      .index = index,
+  };
+  return iree_json_enumerate_object(
+      header_json, iree_io_enumerate_safetensors_entries, &enumerate_state);
+}
+
+IREE_API_EXPORT iree_status_t iree_io_parse_safetensors_index(
+    iree_io_file_handle_t* file_handle, iree_io_parameter_index_t* index) {
+  IREE_ASSERT_ARGUMENT(index);
+  IREE_TRACE_ZONE_BEGIN(z0);
+
+  // Today we only support memory files.
+  // TODO(benvanik): support iree_io_stream_t wrapping for parsing the index.
+  if (iree_io_file_handle_type(file_handle) !=
+      IREE_IO_FILE_HANDLE_TYPE_HOST_ALLOCATION) {
+    IREE_TRACE_ZONE_END(z0);
+    return iree_make_status(IREE_STATUS_UNIMPLEMENTED,
+                            "non-memory safetensors files not yet supported");
+  }
+  iree_byte_span_t host_allocation =
+      iree_io_file_handle_primitive(file_handle).value.host_allocation;
+
+  iree_status_t status = iree_io_parse_safetensors_index_from_memory(
+      file_handle,
+      iree_make_const_byte_span(host_allocation.data,
+                                host_allocation.data_length),
+      index);
+
+  IREE_TRACE_ZONE_END(z0);
+  return status;
+}
```

Comment:
If I had a dollar for every time I'd written something like this... I'd have about $3.23. Good job.